### DISCUSSION TITLE
Pre-computed embeddings from Cohere V2

### DISCUSSION THREAD

**Discussion by rdsch2** [MaartenGr/BERTopic#2116] [Q&A]
I am using BERTopic for topic modelling texts in Tibetan.  The Cohere V2 embeddings have proved excellent for doing semantic searching in Tibetan.  The V3 embeddings are not suitable however (and they use cosine similarity).  But the V2 embeddings are designed for dot-product measures of similarity.  I want to pass a precomputed dot product similarity matrix to  UMAP to reduce dimensionality.  Then pass the umap embeddings to Bertopic.  Is this a reasonable approach.  It does not appear there is any other way to get dot-product optimized embeddings to BERTopic.

**Comment by MaartenGr** [comment]
Yes, you can indeed calculate those embeddings beforehand and then pass the reduced embeddings to BERTopic. You can then simply skip the dimensionality reduction step and go directly to the clustering. Do note though that there are many places where BERTopic uses cosine similarity rather than dot product for things like finding out how similar topics are to one another if you use hierarchical clustering.

**Reply by rdsch2** [reply]
I have been using the Cohere multilingual V2 embeddings successfully for semantic searching.  They recommend using dot product similarity not cosine.  The embeddings retain magnitude as well as direction.  For use with BERTopic I have experimented with normalizing the vectors beforehand and various setting for umap-e.g., metric = cosine, euclidean, etc.  I have found that BERTopic gives me the best results when I just pass the Cohere embeddings as is to BERTopic, which presumably uses UMAP with the default metric of euclidean on my non-normalized embeddings.  Likewise, when I use UMAP with metric=euclidean and pass the embeddings to Plotly to generate a map of the embeddings I get recognizable clusters identified with the topics from BERTopic.  I'm happy with the results, but I don't quite understand why it is working out of the box with the non-normalized embeddings.

**Reply by MaartenGr** [reply]
I can't speak as to how those models were created but considering their newer models also work with cosine similarity, I would not be surprised if their primary focus was dot-product during training but still allowed some training for other metrics, albeit a small part of the process. As a result, it might just be that they simply work better on these metrics than they had initially thought.

**Reply by rdsch2** [reply]
Their newer multilingual model which has a a longer context length and uses cosine similarity works fine with Chinese and I have used that with good results for semantic searching and with BERTopic.  But the V3 embeddings don't work at all with Tibetan.  I've communicated with the company and they have explained that the V3 embeddings were trained differently and recommended that for Tibetan (and I would imagine some other low resource languages) that I stay with V2.  How these models were trained to produce embeddings is something of a black box.
